lib/spack/spack/solver/concretize.lp

```diff
@@ -7,31 +7,114 @@
 % This logic program implements Spack's concretizer
 %=============================================================================
 
+% ID of the nodes in the "root" link-run sub-DAG
+#const main_node_id = 0.
+
+#const link_run = 0.
+#const direct_link_run =1.
+#const direct_build = 2.
+
+% Allow clingo to create nodes
+{ attr("node", node(0..X-1, Package))         } :- max_nodes(Package, X), not virtual(Package).
+{ attr("virtual_node", node(0..X-1, Package)) } :- max_nodes(Package, X), virtual(Package).
```

Comment:
same here -- stick to "dupes" as the term for multiple instances of the same package:

```suggestion
{ attr("node", node(0..X-1, Package))         } :- max_dupes(Package, X), not virtual(Package).
{ attr("virtual_node", node(0..X-1, Package)) } :- max_dupes(Package, X), virtual(Package).
```

Reply:
Done in e716dcef0c3789ae9acd40a388fc2033b860fe62